### PR TITLE
Avoid OutOfMemoryException in some test environments

### LINKED ISSUE #43932: Test failure: System.Buffers.Tests.ArrayBufferWriterTests_Byte.GetMemory_ExceedMaximumBufferSize
failed in job: [ runtime-libraries-coreclr outerloop 20201027.2 ](https://dev.azure.com/dnceng/public/_build/results?buildId=867094&view=ms.vss-test-web.build-test-results-tab&runId=27760178&resultId=104647&paneView=debug)

net6.0-Windows_NT-Release-x64-CoreCLR_release-(Windows.Server.Core.1909.Amd64.Open)windows.10.amd64.server20h1.open@mcr.microsoft.com/dotnet-buildtools/prereqs:windowsservercore-2004-helix-amd64-20200904200251-272704c

Error message
~~~
System.OutOfMemoryException : Exception of type 'System.OutOfMemoryException' was thrown.


Stack trace
   at System.Array.Resize[T](T[]& array, Int32 newSize) in /_/src/libraries/System.Private.CoreLib/src/System/Array.cs:line 68
   at System.Buffers.ArrayBufferWriter`1.CheckAndResizeBuffer(Int32 sizeHint) in /_/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs:line 199
   at System.Buffers.ArrayBufferWriter`1.GetMemory(Int32 sizeHint) in /_/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs:line 133
   at System.Buffers.Tests.ArrayBufferWriterTests_Byte.GetMemory_ExceedMaximumBufferSize() in /_/src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.Byte.cs:line 109
~~~

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/runtime/issues/43932

### CODE REVIEW

File: src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.Byte.cs [hunk 1]
```diff
@@ -106,12 +106,21 @@ public void GetMemory_ExceedMaximumBufferSize()
...
             // Validate we can't double the buffer size, but can grow
-            Memory<byte> memory = output.GetMemory(1);
-
-            // The buffer should grow more than the 1 byte requested otherwise performance will not be usable
-            // between 1GB and 2GB. The current implementation maxes out the buffer size to MaxArrayLength.
-            Assert.Equal(MaxArrayLength - initialCapacity, memory.Length);
-
+            Memory<byte> memory;
+            try
+            {
```
Comment (ahsonkhan) [inline src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.Byte.cs:111]: If allocating ~2 GB is problematic, could ~1 GB also be an issue? In that case, it might be better to put the try/catch around the whole test, including the initialization of ABW and the first call to Advance.

Alternatively, maybe we should just opt to skip the test for the specific constrained devices where 1-2GB allocation is problematic (with some kind of conditional skip attribute). How much RAM is available on the Windows.Nano test machines?

Comment (steveharter) [reply src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.Byte.cs:111]: Perhaps 1GB could cause issues as well; I see other JSON inner loop tests that allocate ~300K but not up to 1GB.

File: src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.Byte.cs [hunk 2]
```diff
@@ -102,16 +102,27 @@ public void GetMemory_ExceedMaximumBufferSize()
...
+                // between 1GB and 2GB. The current implementation maxes out the buffer size to MaxArrayLength.
+                Assert.Equal(MaxArrayLength - initialCapacity, memory.Length);
+            }
+            catch (OutOfMemoryException)
+            {
+                // On memory constrained devices, we can get an OutOfMemoryException, which we can safely ignore.
+            }
+
+            // Validate > MaxArrayLength.
+            output = new ArrayBufferWriter<byte>(1);
```
Comment (ahsonkhan) [inline src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.Byte.cs:125]: This is changing the semantics of the test. We should consider leaving it as is.

File: src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.Byte.cs [hunk 3]
```diff
@@ -102,16 +102,27 @@ public void GetMemory_ExceedMaximumBufferSize()
...
+                output = new ArrayBufferWriter<byte>(initialCapacity);
+                output.Advance(initialCapacity);
+
+                // Validate we can't double the buffer size, but can grow
+                Memory<byte> memory;
+                memory = output.GetMemory(1);
+
+                // The buffer should grow more than the 1 byte requested otherwise performance will not be usable
+                // between 1GB and 2GB. The current implementation maxes out the buffer size to MaxArrayLength.
+                Assert.Equal(MaxArrayLength - initialCapacity, memory.Length);
```
Comment (ahsonkhan) [inline src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.Byte.cs:117]: After successfully setting ABW to init capacity, verify that `output.GetMemory(int.max)` fails.

```suggestion
                Assert.Equal(MaxArrayLength - initialCapacity, memory.Length);
                Assert.Throws<OutOfMemoryException>(() => output.GetMemory(int.MaxValue));
```

File: src/libraries/System.Memory/System.Memory.sln [hunk 4]
```diff
@@ -1,4 +1,8 @@
-﻿Microsoft Visual Studio Solution File, Format Version 12.00
+﻿
```
Comment (ahsonkhan) [inline src/libraries/System.Memory/System.Memory.sln:1]: Revert unrelated changes to this `.sln` file?

**Review by ahsonkhan — APPROVED** [review]
Otherwise, looks good.
